Automattic/wp-calypso
# Add mobile app banners

Resolves: https://github.com/Automattic/wp-calypso/issues/15228 and https://github.com/Automattic/wp-calypso/issues/15229

This component will be shown on mobile devices for Editor, Stats, and Reader pages, or on any other page provided that Notification panel is open. 

### Summary

We are currently using the `userAgent` in order to detect whether we should show the banner. The banner text and images vary depending on page that they are viewed on. In order to determine which variation to show, we are relying on `getSectionName` selector, along with `isNotificationsOpen`. 

I had to add this component to `layout/index.jsx` in order to be able to show it on any page when Notifications panel is open. This also helps with `position: fixed` layout that we need, as we don't run into positioning issues on pages with scrollbar.

Todo:
- [ ] add new deep links when we get them from mobile team, instead of direct links to the app store.

### Testing instructions
1. This banner should not be visible on any page in non-mobile browsers.
2. In order to test it, you can open up DevTools in Chrome and use `Toggle device toolbar`. Along making the screen size smaller, this will also change the `userAgent` which we use to test whether the app banner should be shown. Another option would be to select the `userAgent` manually in Chrome DevTools by following option 2 of [this guide](https://www.technipages.com/google-chrome-change-user-agent-string).
3. Verify that app banner is shown on stats, reader, and editor pages. It should also be shown on any page when notification panel is open.
4. Verify that correct text and icons are shown for each page as described in https://github.com/Automattic/wp-calypso/issues/15228.
5. Verify that impression tracks event has been recorded upon banner rendering.
6. Verify that `Open in app` redirects to app store and that appropriate tracks event is recorded.
7. Verify that `No thanks` closes the banner and that appropriate tracks event is recorded.

**_Note:_** If you've dismissed the widget and want to bring it back for testing in current session, you can run the following line in your console: 
```javascript
dispatch( { type: 'PREFERENCES_SET', key: 'appBannerDismissTimes', value: null } );
```

### Visual changes (stats example):

<img src="https://user-images.githubusercontent.com/1182160/28293508-bdd14fa0-6b54-11e7-94bf-77cec8f282cb.png" width="300px" alt="stats-example" />

**Janitorial changes:**
 - sorted `assets/stylesheets/shared/functions/_z-index.scss` entries by `z-index` value.
